ntoskrnl/ps/job.c

```diff
@@ -86,52 +128,184 @@ PspInitializeJobStructures(VOID)
 }
 
 NTSTATUS
 NTAPI
 PspTerminateJobObject(PEJOB Job,
-    KPROCESSOR_MODE AccessMode,
-    NTSTATUS ExitStatus )
+                      KPROCESSOR_MODE AccessMode,
+                      NTSTATUS ExitStatus)
 {
-    DPRINT("PspTerminateJobObject() is unimplemented!\n");
-    return STATUS_NOT_IMPLEMENTED;
+    PEPROCESS Process;
+    PLIST_ENTRY Entry;
+    NTSTATUS Status;
+
+    DPRINT("PspTerminateJobObject(%p,%x,%x)\n", Job, AccessMode, ExitStatus);
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&Job->JobLock);
+
+    Entry = Job->ProcessListHead.Flink;
+    /* for each process in job process list */
+    while (Entry != Job->ProcessListHead.Flink)
+    {
+        /* fetch process object */
+        Process = CONTAINING_RECORD(Entry, EPROCESS, JobLinks);
+        ObReferenceObject(Process);
+        Entry = Entry->Flink;
+
+        /* process might be getting deleted already */
+        if (ExAcquireRundownProtection(&Process->RundownProtect))
+        {
+            /* terminate process */
+            Status = PsTerminateProcess(Process, ExitStatus);
+            ASSERT(NT_SUCCESS(Status));
+
+            ExReleaseRundownProtection(&Process->RundownProtect);
+        }
+        else
+        {
+            //__debugbreak();
+        }
+
+        /* say good bye and get next one */
+        ObDereferenceObject(Process);
+    }
+
+    /* resume APCs and release lock */
+    ExReleaseResourceAndLeaveCriticalRegion(&Job->JobLock);
+
+    /* notify anyone waiting for the job object that we are done */
+    KeSetEvent(&Job->Event, IO_NO_INCREMENT, FALSE);
+
+    return STATUS_SUCCESS;
 }
 
 VOID
 NTAPI
 PspRemoveProcessFromJob(IN PEPROCESS Process,
                         IN PEJOB Job)
 {
-    /* FIXME */
+    /* This function is called as the Process is destroyed, see ps/kill.c */
+    DPRINT("PspRemoveProcessFromJob(%p,%p)\n", Process, Job);
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&Job->JobLock);
+
+    /* If we are the one that removed the Job object from the process, also unlink the process */
+    if (InterlockedCompareExchangePointer((PVOID)&Process->Job, NULL, Job) == Job)
+    {
+        RemoveEntryList(&Process->JobLinks);
+        ASSERT((Job->ActiveProcesses-1) < Job->ActiveProcesses);
+        Job->ActiveProcesses--;
+    }
+    else
+    {
+        DPRINT1("PspRemoveProcessFromJob(%p,%p) Process not in job\n", Process, Job);
+    }
+
+    ExReleaseResourceAndLeaveCriticalRegion(&Job->JobLock);
 }
 
 VOID
 NTAPI
 PspExitProcessFromJob(IN PEJOB Job,
                       IN PEPROCESS Process)
 {
-    /* FIXME */
+    /* This function is called when the last thread exits, see ps/kill.c */
+    DPRINT("PspExitProcessFromJob(0x%p, 0x%p)\n", Job, Process);
+
+    /* Make sure we are not interrupted */
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&Job->JobLock);
+
+    /* Decrement counters */
+    if (Process->Job == Job)
+    {
+        //ASSERT((Job->ActiveProcesses-1) < Job->ActiveProcesses);
```

Comment:
Why not? I'm guessing PspRemoveProcessFromJob is always called after this anyway?

Reply:
Becuase PspRemoveProcessFromJob already does this.